OSDOCS-9565: configuring AWS Outposts postinstallation (#71263)
openshift/openshift-docs: modules/aws-outposts-requirements-limitations.adoc

// Module included in the following assemblies:
//
// * post_installation_configuration/configuring-aws-outposts.adoc

:_mod-docs-content-type: REFERENCE
[id="aws-outposts-requirements-limitations_{context}"]
= AWS Outposts on {product-title} requirements and limitations

By configuring your {product-title} cluster to accommodate the requirements and limitations of this platform, you can manage the resources on your Outpost similarly to those on a cloud-based AWS cluster.

* To extend an {product-title} cluster on AWS into an Outpost, you must have installed the cluster into an existing VPC.

* {product-title} clusters on AWS include the `gp3-csi` and `gp2-csi` storage classes.
These classes correspond to Amazon Elastic Block Store (EBS) gp3 and gp2 volumes.
{product-title} clusters use the `gp3-csi` storage class by default, but AWS Outposts does not support EBS gp3 volumes.

> 

Reviewer: For the section, we'd better add those extra limitations.

- To create a volume in the Outpost instance, the CSI driver determines the Outpost ARN based on the topology keys stored on the CSINode objects. To ensure that the CSI driver uses the correct topology values, it is necessary to use the `WaitForConsumer` volume binding mode and avoid setting `allowed topologies` on any new storage class created.
- For the outposts remote workers and public workers mixed configuration, the public AWS Elastic Block volume couldn't be attached to the outposts remote workers, the outposts volumes couldn't be attached to public workers as well (the outposts volume logic persistent volume nodeAffinity already has the required selector could avoid it attach to a public worker while the public volume logic persistent volume nodeAffinity doesn't prevent it to be attach to an outposts worker).
cc to @duanwei33

Author: Good point - these conditions were present in the storage topic that I was asked to remove. I will add them in to the limitations topic so we have them stated :+1:

* An Outpost has its own availability zone in an AWS region and has a dedicated subnet.
Edge compute machines deployed into an Outpost must use the Outpost availability zone and subnet.

* AWS Outposts does not support AWS Network Load Balancers or Classic Load Balancers.
To manages Ingress objects for your edge compute resources, you must install the AWS Load Balancer Operator so that you can use AWS Application Load Balancers in the AWS Outposts environment.
If your cluster contains both edge and cloud-based compute instances that share workloads, additional configuration is required.

* To create a volume in the Outpost, the CSI driver requires the Outpost Amazon Resource Name (ARN).
The driver uses the topology keys stored on the `CSINode` objects to determine the Outpost ARN.
To ensure that the driver uses the correct topology values, you must set the volume binding mode to `WaitForConsumer` and avoid setting allowed topologies on any new storage classes that you create.

* When you extend an AWS VPC cluster into an Outpost, you have two types of compute resources.
The Outpost has edge compute nodes, while the VPC has cloud-based compute nodes.

> The public AWS Elastic Block volume cannot attach to Outpost edge compute nodes, and the Outpost volumes cannot attach to public compute nodes.

Author: When I spoke to someone at AWS about this, the corrected "remote" to be "edge" but made no comment on my use of "cloud-based", so I have assumed that terminology is fine. Agree wrt s/public/cloud-based here :+1:

+

> As a result, external tools cannot use CSI snapshots or persistent volume claims to migrate applications from cloud-based compute nodes to edge compute nodes.

Reviewer: ```suggestion
As a result, external tools cannot use CSI snapshots or persistent volume claims to migrate applications which using the persistent storage from cloud-based compute nodes to edge compute nodes.
```
Maybe better, since only applications using the persistent storage will be impacted while not all the applications using the persistent storage.

Author: Thanks so much for helping refine this, @Phaow :smiley:

> To migrate applications, you must perform a manual backup and restore operation.

Reviewer: ```suggestion
To migrate applications persistent storage data, you must perform a manual backup and restore operation.
```
Maybe better.

* An Outpost has its own availability zone in an AWS region and has a dedicated subnet.
Edge compute machines deployed into an Outpost must use the Outpost availability zone and subnet.

* AWS Outposts does not support AWS Network Load Balancers or Classic Load Balancers.
To manages Ingress objects for your edge compute resources, you must install the AWS Load Balancer Operator so that you can use AWS Application Load Balancers in the AWS Outposts environment.
If your cluster contains both edge and cloud-based compute instances that share workloads, additional configuration is required.

* To create a volume in the Outpost, the CSI driver requires the Outpost Amazon Resource Name (ARN).
The driver uses the topology keys stored on the `CSINode` objects to determine the Outpost ARN.
To ensure that the driver uses the correct topology values, you must set the volume binding mode to `WaitForConsumer` and avoid setting allowed topologies on any new storage classes that you create.

* When you extend an AWS VPC cluster into an Outpost, you have two types of compute resources.
The Outpost has edge compute nodes, while the VPC has cloud-based compute nodes.

> The public AWS Elastic Block volume cannot attach to Outpost edge compute nodes, and the Outpost volumes cannot attach to public compute nodes.
> The Outpost volume logic uses node affinity to avoid attaching to public compute nodes, but the public AWS Elastic Block volume logic does not prevent it from attaching to Outpost edge compute nodes.

Reviewer: ```suggestion
The public AWS Elastic Block volume cannot attach to Outposts edge compute nodes, and the Outposts volumes cannot attach to cloud-based compute nodes as well, the limitation also for CSI snapshot restore volumes.
```
The section is introducing the limitation, so we could omit the second one, the second is try to explain why the first limitation exist. The Outpost volume object(PV) uses node affinity to avoid attaching to public compute nodes, but the public AWS Elastic Block volume object does not have node affinity prevent it from attaching to Outpost edge compute nodes. So user should aware this if they want to migrate some apps from cloud-based to edge they needs to consider the persistent volumes need to manual backup and restore(by other external tools couldn't use the csi snapshot or the same pvc across different compute nodes).

Author: Thank you for clarifying this for me! I think the cloud-based to edge migration use case would be helpful to mention here. I will add that so the user can recognize why the limitation is important for them in that case.

// Does the user need to do something to prevent this manually?
// Also, how do we reconcile these two statements?
// "The public AWS Elastic Block volume cannot attach to Outpost edge compute nodes"
// but

> // "the public AWS Elastic Block volume logic does not prevent it from attaching to Outpost edge compute nodes"

Reviewer: Sure, added some more details, sorry for the previous comment doesn't clealy enough.

> // "the public AWS Elastic Block volume logic does not prevent it from attaching to Outpost edge compute nodes"

Author: Thanks so much, and no need to apologize. It's outside my typical domain, so I need a little extra hand-holding in parts of this :slightly_smiling_face:

:_mod-docs-content-type: REFERENCE
[id="aws-outposts-requirements-limitations_{context}"]
= AWS Outposts on {product-title} requirements and limitations

By configuring your {product-title} cluster to accommodate the requirements and limitations of this platform, you can manage the resources on the AWS Outpost similarly to a cloud-based AWS cluster.

* To extend an {product-title} cluster on AWS into an AWS Outpost, you must have installed the cluster into an existing VPC.

* {product-title} clusters on AWS include the `gp3-csi` and `gp2-csi` storage classes.
These classes correspond to Amazon Elastic Block Store (EBS) gp3 and gp2 volumes.
{product-title} clusters use the `gp3-csi` storage class by default, but AWS Outposts does not support EBS gp3 volumes.
You must configure the compute machine set that creates machines in your AWS Outposts instance to use gp2 volumes.
+
For more information about AWS general-purpose EBS volume types, see the AWS documentation link:https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-plan-storage-compare-volume-types.html[Comparing Amazon EBS volume types gp2 and gp3].

////

> //The existing limitation text has these `oc` commands to change the default storage class. Do we need this, or can we just have the machineset use gp2 for machines it creates?

Reviewer 2: I think the procedure is also about persistent volume claims, so we need to cover that somewhere, but, machineset setting gp2 is enough for getting machines up

[source,terminal]
----
$ oc annotate --overwrite storageclass gp3-csi storageclass.kubernetes.io/is-default-class=false
$ oc annotate --overwrite storageclass gp2-csi storageclass.kubernetes.io/is-default-class=true
----
////

////
//Per Joel, inaccurate:
To use AWS Outposts, all compute machine instances must be located within the same Outpost instance and cannot be located in an AWS region.
It is not possible for the cluster to have instances in both AWS Outposts and an AWS region.
//This is the current understanding:
////
* An AWS Outpost has its own availability zone in an AWS region and has a dedicated subnet.
Remote compute machines deployed into an Outpost must use the Outpost availability zone and subnet.

//This statement needs verification, it was part of the note about regions:
* Control plane nodes on a cluster that uses AWS Outposts must not be schedulable.

> //If true, need to explain how to check for and rectify this.

Reviewer 2: I don't think this statement is relevant to this page of the documentation
